curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md

```diff
@@ -0,0 +1,208 @@
+---
+id: 66f3f6eb66ea9dc41cdc30df
+title: Design Colored Boxes
+challengeType: 14
+dashedName: "lab-colored-boxes"
+demoType: onClick
+---
+
+# --description--
+
+In this lab, you'll practice using CSS colors by designing boxes.
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.
+
+**User Stories:**
+
+1. You should set the background color for `body` to `#f4f4f4`.
+2. You should have a `div` with a class of `color-grid` to hold all your color elements.
+3. You should have five `div` elements within `color-grid` `div`.
+4. The five `div` elements should each have a class of `color-box` and `color#`, where `#` is the number of the order of that `div`. For example: `color1` for the first `div`, `color2` for the second, and so on.
+5. The `.color-box` class should have a set `width` and `height` so your `div` elements are visible on the page.
+6. The `.color1` element should have a `background-color` that uses hexadecimal color value.
+7. The `.color2` element should have a `background-color` that uses an RGB color value.
+8. The `.color3` element should have a `background-color` that uses a predefined (word) color value.
+9. The `.color4` element should have a `background-color` that uses a HSL color value.
+10. The `.color5` element should have a `background-color` set.
+
+# --hints--
+
+`body` should have a background color of `#f4f4f4`.
+
+```js
+const body = document.body;
+const bodyBgColor = getComputedStyle(body).backgroundColor;
+assert.strictEqual(bodyBgColor, 'rgb(244, 244, 244)');
+```
+
+You should have a `div` element with a class of `color-grid`.
+
+```js
+const colorGrid = document.querySelector('div.color-grid');
+assert.exists(colorGrid);
+```
+
+You should have five `div` elements within the `.color-grid` `div` element.
+
+```js
+const colorGridChildren = document.querySelectorAll('div.color-grid > div');
+assert.strictEqual(colorGridChildren.length, 5);
+```
+
+Each of the five `div` elements should each have a class of `color-box` and `color#`—substitute the order of the `div` for the `#` symbol.
+
+```js
+const colorGridChildren = document.querySelectorAll('div.color-grid > div');
+assert.strictEqual(colorGridChildren.length, 5);
+
+colorGridChildren.forEach((child, index) => {
+  const colorClass = `color${index + 1}`;
+  assert.isTrue(child.classList.contains('color-box'));
+  assert.isTrue(child.classList.contains(colorClass));
+});
+```
+
+The `.color-box` element should have a set `width` and `height`.
+
+```js
+const colorBox = document.querySelector('.color-box');
+assert.exists(colorBox);
+
+const colorBoxStyles = getComputedStyle(colorBox);
+const width = colorBoxStyles.width;
+const height = colorBoxStyles.height;
+
+assert.notStrictEqual(width, '0px');
+assert.notStrictEqual(height, '0px');
+```
+
+The `.color1` element should have a hexadecimal background color.
+
+```js
+assert.match(code, /\.color1\s*{[^}]*\bbackground-color\s*:\s*#[0-9a-fA-F]{3,6}\s*;[^}]*}/);
```

Comment:
```suggestion
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color1').getPropVal('background-color', true),
  /^#[0-9a-fA-F]{3,6}$/
);
```

wouldn't a test like this avoid the issue of comments, and eventual other properties?

Reply:
@ilenia-magoni, the issue with these formats is how the browser handles/converts different CSS rules. For example:

```js
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color2').getPropVal('background-color', true),
  /^rgb\s*\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*\)$/
);
```

The test above will pass for both rgb and hsl. The tests you suggested for `.color1` and `.color4` don't work, but with a little workaround, `.color1` starts passing for hex, rgb, and hsl values.

IMO, using regex here ensures each test passes for specific rules. I couldn't find a workaround before switching to regex.

CC @moT01 @jdwilkin4